src/main/java/io/odpf/firehose/sink/mongodb/MongoSink.java

```diff
@@ -0,0 +1,103 @@
+    private final MongoClient mongoClient;
+    private List<Message> messages;
+
+    /**
+     * Instantiates a new Mongo sink.
+     *
+     * @param instrumentation               the instrumentation
+     * @param sinkType                      the sink type
+     * @param mongoCollection               the mongo collection
+     * @param mongoClient                   the mongo client
+     * @param mongoRequestHandler           the mongo request handler
+     * @param mongoRetryStatusCodeBlacklist the mongo retry status code blacklist
+     */
+    public MongoSink(Instrumentation instrumentation, String sinkType, MongoCollection<Document> mongoCollection, MongoClient mongoClient, MongoRequestHandler mongoRequestHandler,
+                     List<String> mongoRetryStatusCodeBlacklist) {
+        super(instrumentation, sinkType);
+        this.mongoCollection = mongoCollection;
+        this.mongoRequestHandler = mongoRequestHandler;
+        this.mongoClient = mongoClient;
+        this.mongoRetryStatusCodeBlacklist = mongoRetryStatusCodeBlacklist;
+    }
+
+    @Override
+    protected void prepare(List<Message> messageList) {
+        this.messages = messageList;
+        bulkRequest.clear();
+
+        messages.forEach(message -> bulkRequest.add(mongoRequestHandler.getRequest(message)));
+    }
+
+    @Override
+    protected List<Message> execute() throws Exception {
+        List<BulkWriteError> bulkWriteErrors = processRequest(bulkRequest);
+
+        if (bulkWriteErrors.isEmpty()) {
+            return Collections.emptyList();
+        }
+        getInstrumentation().logWarn("Bulk request failed");
+        handleBulkWriteErrors(bulkWriteErrors);
+
+        List<Message> failedMessages = new ArrayList<>();
+
+        bulkWriteErrors.forEach((bulkWriteError) -> {
+            if (!mongoRetryStatusCodeBlacklist.contains(String.valueOf(bulkWriteError.getCode()))) {
+                failedMessages.add(messages.get(bulkWriteError.getIndex()));
+            }
+        });
+        return failedMessages;
+    }
+
+    @Override
+    public void close() throws IOException {
+        getInstrumentation().logInfo("MongoDB connection closing");
+        this.mongoClient.close();
+    }
+
+    protected List<BulkWriteError> processRequest(List<WriteModel<Document>> bulkRequest) {
+        return MongoResponseHandler.processRequest(bulkRequest, mongoCollection, getInstrumentation());
+    }
+
+    private void handleBulkWriteErrors(List<BulkWriteError> bulkWriteErrors) {
+        int failedResponseCount = 0;
+        for (BulkWriteError bulkWriteError : bulkWriteErrors) {
+            failedResponseCount++;
+            String responseStatus = String.valueOf(bulkWriteError.getCode());
+            if (mongoRetryStatusCodeBlacklist.contains(responseStatus)) {
+                getInstrumentation().logInfo("Not retrying due to response status: {} is under blacklisted status code", responseStatus);
```

Comment:
This message is not valid anymore right? You can put something like logWarn("non-retriable error", status)

Reply:
should i remove the line 96 and replace it with this - logWarn("non-retriable error", status)

Reply:
fixed